src/v/raft/tests/stm_manager_test.cc

```diff
@@ -351,3 +351,134 @@ TEST_F_CORO(
+              "Current batch last offset: {}",
+              _allow_apply,
+              batch.last_offset()));
+        }
+        vassert(
+          batch.base_offset() == next(),
+          "batch {} base offset is not the next to apply, expected base "
+          "offset: {}",
+          batch.header(),
+          next());
+        co_await simple_kv::apply(batch);
+        co_return;
+    }
+
+    ss::future<> apply_raft_snapshot(const iobuf& buffer) override {
+        return simple_kv::apply_raft_snapshot(buffer);
+    };
+
+    void allow_apply_to(model::offset o) { _allow_apply = o; }
+
+    model::offset _allow_apply;
+};
+
+struct controllable_throwing_kv_2 : public controllable_throwing_kv {
+    using controllable_throwing_kv::controllable_throwing_kv;
+
+    static constexpr std::string_view name = "controllable_throwing_kv_2";
+};
+
+struct controllable_throwing_kv_3 : public controllable_throwing_kv {
+    using controllable_throwing_kv::controllable_throwing_kv;
+
+    static constexpr std::string_view name = "controllable_throwing_kv_3";
+};
+TEST_F_CORO(state_machine_fixture, test_all_machines_throw) {
+    /**
+     * This test covers the scenario in which all state machines thrown an
+     * exception during apply, and then one of the state machines makes some
+     * progress.
+     */
+    create_nodes();
+    std::vector<ss::shared_ptr<simple_kv>> stms;
+    for (auto& [id, node] : nodes()) {
+        raft::state_machine_manager_builder builder;
+        auto kv_1 = builder.create_stm<controllable_throwing_kv>(*node);
+        auto kv_2 = builder.create_stm<controllable_throwing_kv_2>(*node);
+        auto kv_3 = builder.create_stm<controllable_throwing_kv_3>(*node);
+
+        stms.push_back(ss::dynamic_pointer_cast<simple_kv>(kv_1));
+        stms.push_back(ss::dynamic_pointer_cast<simple_kv>(kv_2));
+        stms.push_back(ss::dynamic_pointer_cast<simple_kv>(kv_3));
+
+        co_await node->init_and_start(all_vnodes(), std::move(builder));
+    }
+    for (auto& [id, node] : nodes()) {
+        node->raft()
+          ->stm_manager()
+          ->get<controllable_throwing_kv>()
+          ->allow_apply_to(model::offset(100));
+        node->raft()
+          ->stm_manager()
+          ->get<controllable_throwing_kv_2>()
+          ->allow_apply_to(model::offset(100));
+        node->raft()
+          ->stm_manager()
+          ->get<controllable_throwing_kv_3>()
+          ->allow_apply_to(model::offset(150));
+    }
+    vlog(logger().info, "Generating state for test");
+    auto expected = co_await build_random_state(
+      500, wait_for_each_batch::no, 1);
+    vlog(logger().info, "Waiting for state machines");
+    RPTEST_REQUIRE_EVENTUALLY_CORO(15s, [&] {
+        return std::ranges::all_of(
+          nodes() | std::views::values, [&](auto& node) {
+              auto la = node->raft()
+                          ->stm_manager()
+                          ->template get<controllable_throwing_kv_3>()
```

Comment:
nit: I'd rather explicitly specified the lambda argument type TBH to give more context both to reader and to compiler
(but this is rather about our code style in general, so probably out of scope)